### PR TITLE
[dist, ckpt, lora, docs] feat: support outer extra-parallel mesh placement

### PR DESCRIPTION
**Pull request by mo-ke-ke**
### What does this PR do?

VeOmni exposes `ep_outside` and `extra_parallel_placement_innermost`, but `init_parallel_state()` currently rejects every true value. This PR implements the advertised alternate rank ordering for ExtraParallel + FSDP2 while preserving the existing default.

With 32 DP-shard ranks and EP8:

- default: `(ep_fsdp=4, ep=8)`, so EP8 groups are contiguous;
- outside: `(ep=8, ep_fsdp=4)`, so expert-FSDP4 groups are contiguous.

The HSDP replicate dimension remains outermost in both layouts.

The checkpoint path also follows the named mesh dimensions. Model, optimizer, DCP-to-HF, and LoRA state handling no longer assumes that ExtraParallel is the last mesh dimension.

### Checklist Before Starting

- Searched open issues and PRs for `ep_outside`, `extra_parallel_placement_innermost`, and outer ExtraParallel mesh placement; no duplicate implementation was found.
- PR title follows the enforced `[{modules}] {type}: {description}` format.

### Test

Fresh local tests at commit `ed6cc7124e847b626b09fa232b1c3d1271a7d1af`:

```text
pytest -q tests/distributed/test_extra_parallel_mesh_topology.py
4 passed

pytest -q tests/utils/test_extra_parallel_clip_grad_norm.py \
  -k 'inf_norm or expected_outer_extra_parallel_group_ranks'
6 passed, 19 deselected

pytest -q tests/checkpoints/test_dcp_checkpointer.py \
  -k 'not test_save_dtype_only_casts_floating_tensors'
37 passed, 1 deselected, 2 xfailed

pytest -q tests/utils/test_save_safetensor_utils.py
17 passed

make quality
All checks passed; 635 files already formatted
```

The checkpoint unit regressions cover both mesh orders, HSDP placement, named FSDP submesh selection, non-`ep` ExtraParallel names, and LoRA propagation of both EP and FSDP shard axes. The excluded DCP-to-HF dtype test reaches an unrelated existing `torch.cpu.empty_cache` failure on the local macOS CPU environment.

The existing 8-GPU checkpoint suite now includes two `ep_outside=True` end-to-end cases:

- EP4 + plain FSDP2;
- EP2 + HSDP with replicate degree 2.

Both cases train, save and reload model plus optimizer state, merge DCP to HF, and compare the restored states. The MoE-LoRA EP2 integration path also runs with the outside layout for both shared and independent adapters. These GPU cases are collected locally and are wired into the repository's 8-GPU CI; their CI result is intentionally not pre-claimed here.

Single-node 8-GPU FSDP2 training validation was previously run at commit `bbf2d9b19fbf61cc4cdb58159cae6c62e53d4421` on 8 NVIDIA B300 GPUs with PyTorch 2.10.0+cu129 and Transformers 5.9.0:

- `ep_outside=True` constructed `DeviceMesh((ep=4, ep_fsdp=2), stride=(2, 1))`; actual EP and expert-FSDP process-group membership assertions passed.
- `ep_outside=False` constructed the unchanged default `DeviceMesh((ep_fsdp=2, ep=4), stride=(4, 1))`.
- Both layouts completed three forward/backward/gradient-clipping steps; both distributed runs exited 0.

This is single-node correctness evidence, not a cross-node topology or performance benchmark.

### API and Usage Example

No new config field is added. The existing expert-parallel convenience flag now works:

```yaml
train:
  accelerator:
    ep_size: 8
    ep_outside: true
```

The flag changes rank ordering only. It is not an unconditional performance optimization; users should benchmark complete training steps on their target topology.

`extra_parallel_placement_innermost` is retained for compatibility, but its historical name is misleading: `True` selects the outside layout. The CLI help and argument reference now state this explicitly.

### Design & Code Changes

- Add a pure mesh-spec builder that keeps rank-ordering policy independent from process-group construction.
- Remove the unconditional rejection of outer ExtraParallel placement.
- Select FSDP submeshes and checkpoint DTensor placements by dimension name rather than position.
- Preserve the actual EP and FSDP shard axes in model, optimizer, and LoRA checkpoint paths.
- Add CPU topology/checkpoint contract coverage and 8-GPU FSDP2/HSDP/LoRA end-to-end cases.
- Document the legacy placement-field naming, topology tradeoff, and named-dimension invariant.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] Applied repository quality checks.
- [x] Added/updated documentation and agent constraints.
- [x] Added focused CPU regressions and 8-GPU CI coverage.

## Summary by CodeRabbit

* **New Features**
  * Added configurable extra-parallel mesh placement inside or outside FSDP dimensions.
  * Added named mesh-dimension handling, supporting flexible ExtraParallel/FSDP ordering.
  * Added validation for incompatible parallel sizes and clearer configuration guidance.

* **Bug Fixes**
  * Improved checkpoint save/load and LoRA tensor handling across alternate extra-parallel layouts.
  * Preserved HSDP replication behavior across supported mesh configurations.

* **Tests**
  * Expanded coverage for mesh topology, distributed training, gradient clipping, LoRA, and checkpoint round trips.

### CODE REVIEW

File: tests/checkpoints/test_dcp_checkpointer.py [hunk 1]
```diff
@@ -22,6 +22,134 @@
 from veomni.utils.checkpoint_utils import should_skip_hf_weight_load
 
 
+# ---------------------------------------------------------------------------
+# ExtraParallel checkpoint mesh semantics
+# ---------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (mo-ke-ke) [reply tests/checkpoints/test_dcp_checkpointer.py]: Fixed in ed6cc712: removed the newly added decorative section divider and left the test class unchanged.

File: tests/utils/test_extra_parallel_clip_grad_norm.py [hunk 2]
```diff
@@ -170,6 +170,13 @@ def main():
     if ps.extra_parallel_group("emb") and ps.extra_parallel_fsdp_device_mesh["emb"] is not None:
         emb_fsdp_group = ps.extra_parallel_fsdp_device_mesh["emb"]["emb_fsdp"].get_group()
 
+    if args.train.accelerator.ep_outside:
+        rank = dist.get_rank()
+        expected_ep_group = list(range(rank % 2, dist.get_world_size(), 2))
+        expected_ep_fsdp_group = list(range(rank // 2 * 2, rank // 2 * 2 + 2))
+        assert dist.get_process_group_ranks(ep_group) == expected_ep_group
+        assert dist.get_process_group_ranks(ep_fsdp_group) == expected_ep_fsdp_group
```
[the remark this answers is not in the document]
Comment (mo-ke-ke) [reply tests/utils/test_extra_parallel_clip_grad_norm.py]: Fixed in ed6cc712: the assertion now runs only for active EP, derives EP and EP-FSDP rank lists from world size, replicate degree, and EP size, and has focused cases for plain FSDP and HSDP layouts.
